File: 백준/Gold/1717. 집합의 표현/집합의 표현.js
const fs = require("fs");

const reader = (function () {
  const lines = fs
    .readFileSync(process.platform === "linux" ? 0 : "input.txt", "utf-8")
    .split("\n");

  let cursor = 0;

  function get_line() {
    return lines[cursor++];
  }

  function get_integers() {
    return get_line()
      .split(" ")
      .map((token) => parseInt(token));
  }

  return Object.freeze({
    get_integers,
  });
})();

function solve() {
  const [N, M] = reader.get_integers();
  const parents = Array.from({ length: N + 1 }, (_, i) => i);
  const answers = [];

  function union(u, v) {
    const parent_u = find(u);
    const parent_v = find(v);
    parents[parent_u] = parent_v;
  }

  function find(v) {
    if (v !== parents[v]) {
      parents[v] = find(parents[v]);
    }
    return parents[v];
  }

  for (let i = 0; i < M; i++) {
    const [operation, u, v] = reader.get_integers();
    if (operation === 0) {
      union(u, v);
    } else if (operation === 1) {
      answers.push(find(u) === find(v) ? "YES" : "NO");
    }
  }

  return answers.join("\n");
}

console.log(solve());

File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys

input = sys.stdin.readline
print = sys.stdout.write

N, M = map(int, input().split())
A = [x for x in range(N + 1)]

def find(v):
    if A[v] == v:
        return v
    대표노드 = find(A[v])
    A[v] = 대표노드
    return 대표노드

def union(u, v):
    대표노드1 = find(u)
    대표노드2 = find(v)
    A[대표노드2] = 대표노드1

for _ in range(M):
    명령, 노드1, 노드2 = map(int, input().split())

    if 명령 == 0: # union
        union(노드1, 노드2)
    elif 명령 == 1: # find
        if find(노드1) == find(노드2):
            print("YES\n")
        else:
            print("NO\n")
        

